docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +180,239 @@ Cluster Service Administration
+  the local machine, but may also interact with Kubernetes API and Salt API if
+  needed.
+
+
+``metalk8s`` kubectl plugin
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Language
+""""""""
+
+``Go`` is chosen as the language for kubectl plugin for the following reasons:
+
+- Great interactions with Kubernetes API.
+- Often used for operators and kubectl plugins
+  (`Sample CLI plugin <https://github.com/kubernetes/sample-cli-plugin>`_,
+  `Helpers for kubectl plugins <https://github.com/kubernetes/cli-runtime>`_).
+- Easy to ship because it's a statically compiled binary, no deps to provide.
+- Simple deployment (no real requirements), just drop a binary in the PATH.
+
+.. todo::
+
+  - As kubectl plugin may run from outside of the cluster, think about
+    compilation of binaries for non-"classic linux x86_64"
+
+Input and Output
+""""""""""""""""
+
+Each command should follow the kubectl style and standard as much as possible:
+
+- Command style:
+
+  .. code-block:: bash
+
+    kubectl metalk8s <action> <resource>
+
+- Interactive:
+
+  No interaction with the user, except when it's an ``edit`` command an editor
+  is opened (if needed) and when a password is required a prompt appears to ask
+  it.
+
+- Output style:
+
+  Default human-readable output (``<object> <action>ed``).
+
+  A ``--output``, ``-o`` option to change output format, at least support
+  for ``json`` and ``yaml`` (``jsonpath`` and ``go-template`` when it's
+  possible).
+
+  Internally each action result should be an "object"
+  (e.g.: single-level dictionary) containing several informations,
+  at least:
+
+  - name
+  - message
+  - result (True or False)
+  - an elapsed time (to know each action time)
+
+  A ``--verbose``, ``-v`` option to change log level verbosity (default output
+  to ``stderr``), using
+  `Kubernetes log library <https://github.com/kubernetes/klog>`_.
+
+  By default each command will wait for a result but, when it's possible, a
+  ``--async`` option should allow to do not wait for a result and just output
+  an ID (e.g.: Job ID for Salt) that can be used to watch for the result.
+
+SaltAPI interaction
+"""""""""""""""""""
+
+If the plugin needs to access Salt API then it should use the service proxy
+``http://<apiserver_host>/api/v1/namespaces/kube-system/services/https:salt-master:api/proxy/``.
+
+For each and every Salt API call plugin will need authentication on
+apiserver to access the Salt API service proxy and also to Salt API.
+
+.. note::
+
+  Right now, Salt API only accepts authentication using Bearer token, but in
+  kubeconfig we could have certificates authentication so this kind of
+  kubeconfig will not work with this kubectl plugin.
+
+  Add support for certificates based authentication in Salt API look quite hard
+  and costly.
+
+.. todo::
+
+  Check what can be done to support certificates based kubeconfig or if we
+  consider that we only support token-based kubeconfig for our plugin commands
+  interacting with Salt API.
```

Comment:
Do we have a ticket for this one? Would be good to dump the conversation we had in there.

Reply:
No, but I think we should discuss a bit more about this one before creating a ticket